daml/Marketplace/Custody.daml

```diff
@@ -224,3 +258,48 @@ template TransferDepositRequest
     depositCid : ContractId AssetDeposit
   where
     signatory operator, provider, customer
+
+template LifecycleRequest
+  with
+    operator : Party
+    provider : Party
+    customer : Party
+    assetDepositCid: ContractId AssetDeposit
+    postAssetVersion: Int
+  where
+    signatory operator, provider, customer
+
+    -- TODO: this shouldn't be exported; it's only called from CustodyService.Lifecycle
+    choice Lifecycle' : [ ContractId AssetDeposit ]
+      with
+          investor: Party
+          safekeepingDepositCid: ContractId AssetDeposit
+          fixings: Map Text (Map Date Decimal) --TODO these should be looked up from a contract on-ledger, not provided by the custodian!
+      controller investor  
+        do
+          t <- toDateUTC <$> getTime
+          assetDeposit <- fetch assetDepositCid
+          (_, description) <- fetchByKey @AssetDescription assetDeposit.asset.id
+          (_, description') <- fetchByKey @AssetDescription (assetDeposit.asset.id with version = postAssetVersion)
+          let payout: Id -> Decimal -> Date -> Update (ContractId AssetDeposit)
+              payout asset qty dt = do
+                payoutDepositCid :: _ <- exercise safekeepingDepositCid $ AssetDeposit_Split [qty]
+                exerciseByKey @AssetSettlementRule assetDeposit.account.id $
+                   AssetSettlement_AddController description.issuer
+                cid <- exerciseByKey @AssetSettlementRule description.safekeepingAccountId (
+                  AssetSettlement_Transfer with
+                    receiverAccountId = assetDeposit.account.id
+                    depositCid = payoutDepositCid)
+                exerciseByKey @AssetSettlementRule assetDeposit.account.id $
+                  AssetSettlement_RemoveController description.issuer
+                return cid
+              spot : Text -> Date -> Update Decimal
+              spot label dt = pure $ fromSomeNote
+                ("Lifecycle: Underlying data '" <> label <> "' missing on " <> Map.keyToText dt)
+                (Map.lookup label fixings >>= Map.lookup dt)
+              election : Claim Observation Date Id -> Claim Observation Date Id -> Claim Observation Date Id
+              election _ _ = deserialize description'.claims -- will throw if the claims don't match
+          payoutCids <- ContingentClaims.settle payout spot election (deserialize description.claims) t
+          assetDepositCid' <- create assetDeposit with asset = (assetDeposit.asset with id = (assetDeposit.asset.id with version = postAssetVersion))
```

Comment:
Should you not be calling `AssetSettlement_Credit` here instead of creating an asset `on the fly` ?

Reply:
I'm not sure - what's the difference in this case? Let's discuss on zoom.

Reply:
Same goes for archiving the old asset below - should I debit instead?